consulresolver/src/main/java/io/datanerds/grpc/ConsulNameResolver.java

```diff
@@ -0,0 +1,116 @@
+package io.datanerds.grpc;
+
+import com.google.common.base.Preconditions;
+import com.google.common.net.HostAndPort;
+import com.orbitz.consul.Consul;
+import com.orbitz.consul.HealthClient;
+import com.orbitz.consul.cache.ServiceHealthCache;
+import com.orbitz.consul.cache.ServiceHealthKey;
+import com.orbitz.consul.model.ConsulResponse;
+import com.orbitz.consul.model.health.ServiceHealth;
+import com.orbitz.consul.option.ImmutableCatalogOptions;
+import io.grpc.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public class ConsulNameResolver extends NameResolver {
+    public static final String DEFAULT_ADDRESS = "localhost";
+    public static final Integer DEFAULT_PORT = 8383;
+    private static final Logger logger = LoggerFactory.getLogger(ConsulNameResolver.class);
+    private final String authority;
+    private final String service;
+    @GuardedBy("this")
+    private Listener listener;
+    private Consul consul;
+    @GuardedBy("this")
+    private HealthClient healthClient;
+    private Optional<String> datacenter;
+    private Optional<List<String>> tags;
+
+    public ConsulNameResolver(ConsulQueryParameter parameter) {
+        this.authority = parameter.consulAddress.orElse(DEFAULT_ADDRESS + ":" + DEFAULT_PORT); //todo move to URI
+        this.service = parameter.service;
+        this.datacenter = parameter.datacenter;
+        this.tags = parameter.tags;
+
+        HostAndPort consulHostAndPort = HostAndPort.fromString(this.authority).withDefaultPort(DEFAULT_PORT);
+        this.consul = Consul.builder().withHostAndPort(consulHostAndPort).build();
+        this.healthClient = this.consul.healthClient();
+    }
+
+    @Override
+    public String getServiceAuthority() {
+        return this.authority;
+    }
+
+    protected ImmutableCatalogOptions buildCatalogOptions() {
+        ImmutableCatalogOptions.Builder options = ImmutableCatalogOptions.builder();
+
+        if (this.datacenter.isPresent()) {
+            options.datacenter(com.google.common.base.Optional.of(this.datacenter.get()));
+        }
+
+        if (this.tags.isPresent()) {
+            this.tags.get().forEach(options::tag);
+        }
+        return options.build();
+    }
+
+    @Override
+    public void start(Listener listener) {
+        Preconditions.checkState(this.listener == null, "already started");
+        this.listener = Preconditions.checkNotNull(listener, "listener");
```

Comment:
Can that even happen when `checkState` is not throwing an exception? Would probably use `Objects.requireNonNull(T, String)`